4ebula/react-2024Q3
# Hooks and routing

1. Task: https://github.com/rolling-scopes-school/tasks/blob/master/react/modules/tasks/routing.md
2. Screenshot: ![image](https://github.com/user-attachments/assets/488ac95e-2a08-4364-a768-9f7f21bd2937)
3. Done: 14.07.2024 Deadline: 15.07.2024
4. Score: 45/100
- [x] Create a separate branch for this task from the previous task's branch. Branch name: "hooks-and-routing".
- [x] All components must be changed to **functional components**, except **Error Boundary** components, as error boundaries in React still need to be class components.
- [x] All logic should be split into components:  
   - If you need an access either to the component's lifecycle or the state **use hooks**.
   - All data should be stored in the **component's state**.
- [x] Add routing to your application using **React Router**.
- [x] Add a 404 page when user navigates to non-existing route.
- [x] Add and configure a test runner: Jest or Vitest. Test runner should show the test coverage. You should aim to reach at least 80% of the test coverage.
- [x] Add a testing library: React Testing Library. You should add tests for the several scenarios keeping in mind that mocked data should be used instead of real API calls.
- [x] Tests for the Card List component:
   - Verify that the component renders the specified number of cards;
   - Check that an appropriate message is displayed if no cards are present.
- [x] Tests for the Card component:
   - Ensure that the card component renders the relevant card data;
   - Validate that clicking on a card opens a detailed card component;
   - Check that clicking triggers an additional API call to fetch detailed information.
- [x] Tests for the Detailed Card component:  
  - Check that a loading indicator is displayed while fetching data;
  - Make sure the detailed card component correctly displays the detailed card data;
  - Ensure that clicking the close button hides the component.
- [ ] Tests for the Pagination component:
  - Make sure the component updates URL query parameter when page changes.
- [x] Tests for the Search component:  
  - Verify that clicking the Search button saves the entered value to the local storage;
  - Check that the component retrieves the value from the local storage upon mounting.
- [x] Lastly, update Husky to run tests on the pre-push hook, ensuring that tests are automatically executed before any code is pushed.

## Application Requirements

- [ ] Add pagination:
   - implement pagination for your existing item list
   - display the current page in the browser URL using query parameters (e.g. ?page=2, e.g /search/2).
   - the pagination component should appear after receiving the list of all items.
- [ ] Main page displays search results. On item click page should be split into 2 section:
   - left section will continue to display search results;
   - right section should display details using Router Outlet (show loading indicator while making an additional call for details, add control for closing the section, also section should be closed when user clicks on the left section)
   - reflect in the url that "Details" section has been opened for the selected item (e.g. /?frontpage=2&details=1).

## Points

- [x] Tests are added to the project, can be run via a `test` command in the package.json, no errors are produced, test coverage >= 80% (at least for `statements`) - **30**
- [x] Conversion of class components to functional components with hooks - **15**
- [ ] Custom hook to restore search query from LS and put it back to LS on unmount - **5**
- [ ] Pagination is present in both URL and on the page - **20**
- [ ] Upon clicking, open details panel on the right side of the page: - **5**
  - use router outlet, left side of the page should continue displaying the list of results - **10**
  - initiate an additional API call, display a loader, and update the URL - **5**
  - Details panel should be closed either on the "close" button click or on the main panel click - **10**
